admin-dev/themes/new-theme/js/components/pagination/dynamic-grid-paginator.js

```diff
@@ -0,0 +1,246 @@
+  }
+
+  getPagesCount() {
+    const limit = this.getLimit();
+    const total = this.getTotal();
+
+    return Math.ceil(total / limit);
+  }
+
+  getTotal() {
+    return this.$paginationContainer.data('total');
+  }
+
+  validatePageNumber(page) {
+    if (page > this.getPagesCount()) {
+      return this.getPagesCount();
+    }
+
+    if (page < 1) {
+      return 1;
+    }
+
+    return page;
+  }
+
+  extractPageFromDataset(target) {
+    return Number($(target).data('page'));
+  }
+
+  setCurrentPageInput(page) {
+    $(this.selectorsMap.jumpToPageInput).val(page);
+  }
+
+  updatePaginatorForFirstPage() {
+    this.toggleFirstPageAvailability(false);
+    this.togglePreviousPageAvailability(false);
+    this.updateNextPageBtnData(2);
+    this.toggleNextPageAvailability(true);
+    this.toggleLastPageAvailability(true);
+  }
+
+  updatePaginatorForLastPage() {
+    this.toggleFirstPageAvailability(true);
+    this.toggleNextPageAvailability(false);
+    this.toggleLastPageAvailability(false);
+    this.updatePreviousPageBtnData(this.getPagesCount() - 1);
+    this.togglePreviousPageAvailability(true);
+  }
+
+  updatePaginatorForMiddlePage(page) {
+    this.updatePreviousPageBtnData(page - 1);
+    this.updateNextPageBtnData(page + 1);
+    this.toggleFirstPageAvailability(true);
+    this.togglePreviousPageAvailability(true);
+    this.toggleNextPageAvailability(true);
+    this.toggleLastPageAvailability(true);
+  }
+
+  updatePreviousPageBtnData(page) {
+    const previousPageBtn = this.$paginationContainer.find(this.selectorsMap.previousPageBtn);
+    previousPageBtn.data('url', this.router.generate(this.routing.route, this.mergeParamsToKeep({
+      [this.routing.pageKey]: page,
+      [this.routing.limitKey]: this.getLimit(),
+    })));
+    previousPageBtn.data('page', page);
+  }
+
+  updateNextPageBtnData(page) {
+    const nextPageBtn = this.$paginationContainer.find(this.selectorsMap.nextPageBtn);
+    nextPageBtn.data('url', this.router.generate(this.routing.route, this.mergeParamsToKeep({
+      [this.routing.pageKey]: page,
+      [this.routing.limitKey]: this.getLimit(),
+    })));
+    nextPageBtn.data('page', page);
+  }
+
+  toggleFirstPageAvailability(enable) {
+    const firstPageItem = this.$paginationContainer.find(this.selectorsMap.firstPageItem);
+    this.toggleTargetAvailability(firstPageItem, enable);
+  }
+
+  toggleLastPageAvailability(enable) {
+    const lastPageItem = this.$paginationContainer.find(this.selectorsMap.lastPageItem);
+    this.toggleTargetAvailability(lastPageItem, enable);
+  }
+
+  togglePreviousPageAvailability(enable) {
+    const previousPageItem = this.$paginationContainer.find(this.selectorsMap.previousPageItem);
+    this.toggleTargetAvailability(previousPageItem, enable);
+  }
+
+  toggleNextPageAvailability(enable) {
+    const nextPageItem = this.$paginationContainer.find(this.selectorsMap.nextPageItem);
+    this.toggleTargetAvailability(nextPageItem, enable);
+  }
```

Comment:
I've the feeling that this is way too verbose, you'll be able to do the same with only one method with the selector as a param

Reply:
Sorry, Its first time I see someone complain about "too verbose code". Maybe we need more opinions, but I don't agree with you :stuck_out_tongue_closed_eyes:

Reply:
Altough I've managed to use one function instead of many where it made sense (like refreshButtonsData)

Reply:
To me, it doesn't make sense to repeat multiple methods named differently, where your object mapper is verbose enough to understand the code, having 4 methods is making a lot of noises while you read `togglePageAvailability(enable, this.selectorsMap.firstPageItem)` is clear enough to know that you want to toggle the first page availability

Verbose code and readability is all about balancing to me, of course It's my own opinion

Reply:
managed to balance it with Jonathan's suggestion quite nicely I think :+1: